rfd/0112-DB-credential-providers.md

```diff
@@ -0,0 +1,71 @@
+---
+authors: Sai Sandeep Rangisetti (sandeep.sai@flipkart.com)
+state: draft
+---
+
+# RFD 112 - Use a credential provider for DBs
+
+## What
+As of now all the self-hosted databases require enabling certificate authentication in the backend. This RFD proposes
+a way to avoid that requirement. Based on the labels we can decide to fetch credentials from a web service instead of
+using certificates for authentication. When this label is applied we will consider username as a token and exchange it
+for credentials with an external credential provider
+
+## Why
+Enabling certificate authentication requires config update and restart of the mysql server. This will not be scalable in
+large organizations which will have large number of database clusters. Instead, orgs can have a credential provider
+service like Vault which can give the credentials on need basis based on username and database. 
+
+## Detail
+### Overview 
+```
++--------------+               +--------------------+               +------------------+
+|              | pass token as |                    | pass exhanged |                  |
+|     user     +-------------->|      teleport      +-------------->|     database     |
+|              |   username    |                    |  credentials  |                  |
++--------------+               +--------+-----------+               +------------------+
+                                        |  ^
+                                        |  |
+                                        |  | Exchange token with credentials
+                                        V  |
+                               +-----------+--------+
+                               |                    |
+                               |     credential     |
+                               |      provider      |
+                               |                    |
+                               +--------------------+
+
+```
+
+### Get the details of the credential provider
+To exchange the username/token with actual credentials we need to read the required details like url, authentication
+mechanism and other details. We can read the details from the config file like below 
+```yaml
+db_service:
+  token:
+    url_tmpl: http://localhost:9000/database/{{.DBName}}/user/{{.Username}}/token/{{.Token}}
+    connection_timeout: 1s
+    read_timeout: 1s
+    tls:
+      insecure: false
+      # ca path is required if server is cert is self signed
+      ca_path: /tmp/ca.pem
+      # cert and key path required if authentication is based on mutual tls
+      cert_path: /tmp/cert.pem
+      key_path: /tmp/key.pem
+    authentication:
+      # scheme could be HEADER or NONE if it is header we will send fixed set of headers
+      scheme: HEADER
+      headers:
+        X-API-Key: knkncsdjknkn
+```
```

Comment:
What will be the structure of the response from the Credential Provider ? Are we going to support multiple possible response structures (if so, this will need to be configured)?

Reply:
We will be supporting only one response structure. Added sample request and response structures for that

Reply:
What service does this response format come from ? Is this Hashicorp Vault ? Ideally this RFD would identify the common upstream credential providers that it intends to be compatible with - we'll want to make sure anything that's merged into the core of Teleport will be useful to a variety of customers.